cdi/src/main/java/brave/webmvc/TracingProducer.java

```diff
@@ -0,0 +1,101 @@
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.spi.CDI;
+
+import brave.Tracing;
+import brave.context.log4j2.ThreadContextScopeDecorator;
+import brave.propagation.B3Propagation;
+import brave.propagation.ThreadLocalCurrentTraceContext;
+import brave.sampler.Sampler;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import zipkin2.Span;
+import zipkin2.codec.Encoding;
+import zipkin2.reporter.AsyncReporter;
+import zipkin2.reporter.Reporter;
+import zipkin2.reporter.okhttp3.OkHttpSender;
+
+@ApplicationScoped
+public class TracingProducer {
+
+    private static final Logger log = Logger.getLogger(TracingProducer.class.getName());
+
+    /**
+     * @return configured brave tracing
+     */
+    public static Tracing createTracing() {
+        return CDI.current().select(Tracing.class).get();
+    }
+
+    /**
+     * {@link Dependent} scope is used. {@link ApplicationScoped} does not work because
+     * {@link brave.Tracing#clock(brave.propagation.TraceContext)} is a {@code final} method.
+     *
+     * @param spanReporter injected reporter
+     * @param enabled value of {@link ConfigKeys#TRACING_ENABLED}
+     * @param localServiceName value of {@link ConfigKeys#TRACING_NAME}
+     *
+     * @return the configured brave tracing
+     */
+    @Produces
+    @Dependent
+    static Tracing createTracing(final Reporter<Span> spanReporter,
+                                 @ConfigProperty(name = ConfigKeys.TRACING_ENABLED) boolean enabled,
+                                 @ConfigProperty(name = ConfigKeys.TRACING_NAME) String localServiceName) {
+
+        final Tracing.Builder builder = Tracing.newBuilder();
+
+        if (!enabled) {
+            final Tracing noOpTracing = builder.build();
+            noOpTracing.setNoop(true);
+            return noOpTracing;
+        }
+
+        if (null != localServiceName && localServiceName.length() > 0) {
+            log.log(Level.FINE, "Setting localServiceName to: " + localServiceName);
+            builder.localServiceName(localServiceName);
+        }
+
+        if (null != spanReporter) {
+            log.log(Level.FINE, "Setting span reporter to: " + spanReporter);
+            builder.spanReporter(spanReporter);
+        }
+
+        return builder
+                .sampler(Sampler.ALWAYS_SAMPLE)
+                .trackOrphans()
+                .propagationFactory(B3Propagation.FACTORY)
+                .currentTraceContext(
+                        ThreadLocalCurrentTraceContext.newBuilder()
+                                .addScopeDecorator(ThreadContextScopeDecorator.create())
+                                .build())
+                .build();
+    }
+
+    @Produces
+    @Dependent
+    static Reporter<zipkin2.Span> zipkinReporter(
+            @ConfigProperty(name = ConfigKeys.TRACING_REPORTER_ENABLED) final boolean reporterEnabled,
```

Comment:
suggestion would be DRY only one setting. is there a URL or not? actually to make this fit other examples, default the url and don't add an option to disable it (real life it wouldn't be like this, rather a choice of senders ex kafka etc)

Reply:
Good point, removing the additional enabled flag.
How would you do in real-life? Sender is a class from Zipkin, though?!

Reply:
in spring what happens is that each sender is conditional on something. http is default and has a default for localhost Ex spring.zipkin.baseUrl 

there's an override indicated by a different property spring.zipkin.sender.type like spring.zipkin.sender.type=activemq

it is all here https://github.com/spring-cloud/spring-cloud-sleuth/tree/34f61908fde002ce8415c3fbefb2a70567746cd4/spring-cloud-sleuth-zipkin/src/main/java/org/springframework/cloud/sleuth/zipkin2

however, for an example, just use http :D

Reply:
but yeah. there's `@ConditionalOnMissingBean` in spring boot. if there's an analog for that in CDI, that's best as the user can override the sender with whatever they want using java.

Reply:
This is usually done via `@Produces @Alternative`, or, if you need to evaluate a config property, do so in the producer itself. I'm not sure if there is another way. Basically it would be a factory depending on the configured value.